src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                {
+                    return false;
+                }
+
+                return true;
+            }
+
+            private sealed class BlockAnalyzer
+            {
+                private readonly SymbolStartAnalyzer _symbolStartAnalyzer;
+                private readonly Options _options;
+                private bool _hasDelegateCreationOrAnonymousFunction;
+                private bool _escapedDelegate;
+                private bool _hasConversionFromDelegateTypeToNonDelegateType;
+
+                private BlockAnalyzer(SymbolStartAnalyzer symbolStartAnalyzer, Options options)
+                {
+                    _symbolStartAnalyzer = symbolStartAnalyzer;
+                    _options = options;
+                }
+
+                public static void Analyze(OperationBlockStartAnalysisContext context, SymbolStartAnalyzer symbolStartAnalyzer)
+                {
+                    if (HasSyntaxErrors())
+                    {
+                        return;
+                    }
+
+                    // All operation blocks for a symbol belong to the same tree.
+                    var firstBlock = context.OperationBlocks[0];
+                    var supportsDiscard = symbolStartAnalyzer._supportsDiscard(firstBlock.Syntax.SyntaxTree);
+                    var unusedValueExpressionStatementOption = symbolStartAnalyzer._unusedValueExpressionStatementOption;
+                    var unusedValueAssignmentOption = symbolStartAnalyzer._unusedValueAssignmentOption;
+                    if (!TryGetOptions(firstBlock.Syntax.SyntaxTree, firstBlock.Language,
+                                       context.Options, supportsDiscard,
+                                       unusedValueExpressionStatementOption, unusedValueAssignmentOption,
+                                       context.CancellationToken, out var options))
+                    {
+                        return;
+                    }
+
+                    var blockAnalyzer = new BlockAnalyzer(symbolStartAnalyzer, options);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeExpressionStatement, OperationKind.ExpressionStatement);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeDelegateCreationOrAnonymousFunction, OperationKind.DelegateCreation, OperationKind.AnonymousFunction);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeConversion, OperationKind.Conversion);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeFieldOrPropertyReference, OperationKind.FieldReference, OperationKind.PropertyReference);
+                    context.RegisterOperationBlockEndAction(blockAnalyzer.AnalyzeOperationBlockEnd);
+
+                    return;
+
+                    // Local Functions.
+                    bool HasSyntaxErrors()
+                    {
+                        foreach (var operationBlock in context.OperationBlocks)
```

Comment:
can you give me a brief rundown on OperaitonBlocks?  When do i have many of them?  what do they each correspond to?  I have an idea, but i'd like to know from you before i make too many assumptions.

Reply:
Operation blocks are essentially operations with null parent node that are associated to the given context's owning or containing symbol. So for a method, this includes the operation for the primary method body, the constructor initializer (if any), and one for each parameter initializer. For fields/properties, there is a single possible operation block for the initializer.

---
In reply to: [230605149](https://github.com/dotnet/roslyn/pull/30777#discussion_r230605149) [](ancestors = 230605149)

Reply:
Gotcha.  Super cool. Adidtional clarifying questions:

1. how do you know which each corresponds to? i.e. how do you know it's the parameter initializer?  how do you know it's the constructor-initializer?
2. how are attributes represented?
3. can you get OperationBlocks for an INamedTypeSymbol?  or only for things like IMethodSymbol/IFieldSymbol/IPropertySymbol.
4. for an IPropertySymbol how are operations exposed?  Are they always through the accessor methods?  or can they be through the prop symbol itself. i.e.  `int Foo => 0`?

Thanks!

Reply:
> 1. how do you know which each corresponds to? i.e. how do you know it's the parameter initializer?  how do you know it's the constructor-initializer?

Currently, the only way is to check the operation kind. Ideally, we should not stub param initializers in method body's operation block callback, but that would now be a breaking change.

> 2. how are attributes represented?

There are operation nodes with OperationKind.None with childred nodes having the relevant operations (such as say field or method reference). I think the topmost child is an invocation operation, but need to double check.

> 3. can you get OperationBlocks for an INamedTypeSymbol?  or only for things like IMethodSymbol/IFieldSymbol/IPropertySymbol.

Not with the current APIs. I think there have been requests to expose this as a public API instead of black box in the analyzer driver.

> 4. for an IPropertySymbol how are operations exposed?  Are they always through the accessor methods?  or can they be through the prop symbol itself. i.e.  int Foo => 0?

Property initializers are exposed with operation block callbacks with property symbol as context's owning symbol. See [http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.IOperation.UnitTests/IOperation/IOperationTests_ISymbolInitializer.cs,99](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.IOperation.UnitTests/IOperation/IOperationTests_ISymbolInitializer.cs,99). For arrow expression case above, I believe they are exposed as operation blocks on the getter (can't find the unit test right now).

---
In reply to: [230942599](https://github.com/dotnet/roslyn/pull/30777#discussion_r230942599) [](ancestors = 230942599)